securedrop/journalist_app/api.py

```diff
@@ -279,6 +280,26 @@ def get_current_user():
         user = get_user_object(request)
         return jsonify(user.to_json()), 200
 
+    @api.route('/account/new-password', methods=['POST'])
+    @token_required
+    def new_password():
+        user = get_user_object(request)
+        new_password = json.loads(request.data)['new_password']
+
+        try:
+            user.set_password(new_password)
```

Comment:
This is dangerous because if a client doesn't finish the read (broken connection), then the client could be locked out. Changing a passphrase should require multiple steps.

1. Client requests new passphrase
2. Server check auth, sends new passphrase, and stores `new_passphrase_hash` or something in the DB
3. Client replies with the new passphrase to confirm (separate endpoint)
4. Server checks the request matches the new hash them moves it over in the db
5. Server responds that the passphrase has changed

Additionally, errors in the client (failure to display passphrase, crash, etc) could mean that at the network level, everything went just fine, but the user is still locked out.

Because this is happening over Tor, we have to plan for exceptional robustness.

Reply:
redshiftzero:
> 3. server generates a diceware passphrase and provides the user a new_passphrase in the response

heartsucker:
> 2. Server check auth, sends new passphrase, and stores new_passphrase_hash or something in the DB

So, currently, the web interface basically offers a new password suggestion. It's still up to the client to send a new password, which then gets validated (meeting criteria, not against the suggested password). It doesn't get stored in the database.